File: ff-pwa/test_scraper.js
const cloudscraper = require('cloudscraper');
const cheerio = require('cheerio');

const testUrl = 'https://www.forexfactory.com/calendar/78-us-cpi-mm';

console.log('Fetching:', testUrl);

cloudscraper.get(testUrl).then(html => {
    const $ = cheerio.load(html);
    const rows = [];

    // Select the history table rows
    $('.calendar__table tr').each((i, el) => {
        const row = $(el);
        if (row.find('.calendar__actual').length > 0) {
            const date = row.find('.calendar__date').text().trim();
            const actual = row.find('.calendar__actual').text().trim();
            const forecast = row.find('.calendar__forecast').text().trim();
            const previous = row.find('.calendar__previous').text().trim();

            if (date) {
                rows.push({ date, actual, forecast, previous });
            }
        }
    });

    console.log('Found', rows.length, 'historical rows.');
    if (rows.length > 0) {
        console.log('Sample Row:', rows[0]);
    } else {
        console.log('No data found. Checking if HTML contains expected selectors...');
        console.log('HTML snippet:', html.substring(0, 500));
    }
}).catch(err => {
    console.error('Scraping failed:', err.message);
});

File: ff-pwa/test_scraper.py
import urllib.request
from bs4 import BeautifulSoup
import re

def clean_value(val_str):
    if not val_str: return ""
    return val_str.strip()

def scrape_forex_history(url):
    print(f"Testing URL: {url}")
    headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36'}
    req = urllib.request.Request(url, headers=headers)
    
    try:
        with urllib.request.urlopen(req, timeout=15) as response:
            html = response.read().decode('utf-8')
            print(f"Page size: {len(html)} characters")
            
        soup = BeautifulSoup(html, 'html.parser')
        
        # Check for common blocking indicators
        if "Cloudflare" in soup.title.string if soup.title else "":
            print("BLOCKED BY CLOUDFLARE")
            return []

        table = soup.find('table', class_='calendar__table')
        if not table:
            print("No .calendar__table found")
            # Print a bit of the body to see what we got
            print("Body snippet:", soup.body.get_text()[:500] if soup.body else "No body")
            return []
            
        rows = table.find_all('tr', class_='calendar__row')
        print(f"Found {len(rows)} potential rows")
        
        history_list = []
        for row in rows:
            date_td = row.find('td', class_='calendar__date')
            actual_td = row.find('td', class_='calendar__actual')
            
            if date_td and actual_td:
                date_text = date_td.get_text(separator=" ", strip=True) 
                act_text = actual_td.get_text(separator=" ", strip=True)
                
                # Regex to match "Feb 13, 2026" or similar
                date_match = re.search(r'([A-Z][a-z]{2})\s+(\d{1,2}),\s+(\d{4})', date_text)
                if date_match and act_text:
                    clean_date = f"{date_match.group(1)} {date_match.group(2)}, {date_match.group(3)}"
                    print(f"Matched: {clean_date} | Actual: {act_text}")
                    history_list.append({"date": clean_date, "actual": act_text})
        
        return history_list

    except Exception as e:
        print(f"Error: {e}")
        return []

if __name__ == "__main__":
    # Test with Indicator 5
    test_url = "https://www.forexfactory.com/calendar/90-us-import-prices-mm"
    results = scrape_forex_history(test_url)
    print(f"\nTotal results found: {len(results)}")
